KAFKA-7283: Enable lazy mmap on index files and skip sanity check for segments below recovery point

hzxa21 opened the pull request:
Per the KIP-263 discussion, we think we can improve broker restart time by avoiding performing costly disk operations when sanity checking index files for segments below recovery point on broker startup.

This PR includes the following changes:
1. Mmap the index file and populate fields of the index file on-demand rather than performing costly disk operations when creating the index object on broker startup.
2. Skip sanity checks on the time index and offset index of segments.
   1) For segment with offset below the flushed point (recovery point), these segments are safely flushed so we don't need to sanity check the index files. if there are indeed data corruption on disk, given that we don't sanity check the segment file, sanity checking only the indexes adds little benefit.
   2) For segment with offset above the flushed point (recovery point), we will recover these segments in `recoveryLog()` (Log.scala) in any case so sanity checking the index files for these segments is redundant.

We did experiments on a cluster with 15 brokers, each of which has ~3k segments (and there are 31.8k partitions with RF=3 which are evenly distributed across brokers; total bytes-in-rate is around 400 MBps). The results show that rolling bounce time reduces from 135 minutes to 55 minutes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

hzxa21 commented:
@lindong28 Dong, could you help to review this?

dhruvilshah3 commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -34,16 +34,37 @@ import scala.math.ceil
+
+  def file_=(f: File) {
+    _file = f
+  }
```
nit: I generally prefer listing member variables before methods

lindong28 replied:
Since this is more of a code style question and @hzxa21 may not be familiar with that, I will provide some comment here. So the question here is that should we put listing member variables together at the beginning of the class, or should we put each member variable close to its getter/setter.

It seems that the existing code style in Kafka puts the member variables together at the beginning of the class. This is the case for `highWatermarkMetadata`, `_logStartOffset` and `_lastCaughtUpTimeMs` in `Replica.scala`. The benefit of this approach is that developer can easily see all states (i.e. member variable) of the class at the beginning of the file instead of having to look through the entire file to understand the states. On the other hand, even if we don't put member variable close to its getter/setter, developer can still easily know the member variable associated with each getter/setter by just looking at the method name etc.

ijuma replied:
vals/vars should be at the top since the order matters. It's hard to reason about it if they are scattered all over the file. Methods (`defs`) don't have this issue.

hzxa21 replied:
Fixed.

dhruvilshah3 commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -120,45 +145,84 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
+        if (!initialized)
+          initializeMmapAndLength
+      }
+    }
```
Because this block is repeated a few times, could we fold it into a method?

hzxa21 replied:
Done.

dhruvilshah3 commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -34,16 +34,37 @@ import scala.math.ceil
+    _file.getName
+  }
+
+  def file_=(f: File) {
```
Hmm, this looks like something we already do but I wonder if it safe to switch to using a new `File` at any point in time? What happens to other state like `_length` and `_mmap` after the switch?

hzxa21 replied:
That is a good point. I dug into the codes and found that we use the setter of `file` only when we rename the file (e.g. alter log dir, add suffix). In this case, only the metadata (i.e. inode) changes and the physical location of the file stays the same so I think the state like `_length` and `_map` are the same and we don't need to reinitialize them after the switch.

dhruvilshah3 commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -392,18 +392,27 @@ class Log(@volatile var dir: File,
+            // Only sanity check segments above the recovery point
+            segment.sanityCheck()
+        }
+        catch {
```
nit: you could start this `catch` block on the previous line

hzxa21 replied:
Done.

dhruvilshah3 commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -94,8 +88,23 @@ class LogSegment private[log] (val log: FileRecords,
   /* The maximum timestamp we see so far */
-  @volatile private var maxTimestampSoFar: Long = timeIndex.lastEntry.timestamp
-  @volatile private var offsetOfMaxTimestamp: Long = timeIndex.lastEntry.offset
+  @volatile private var _maxTimestampSoFar: Long = -1L
```
Should we use an `Option` instead?

hzxa21 replied:
I don't think `Option` can provide the atomicity guarantee without using a lock, so I think it is safer to just keep `volatile` Long as before.

dhruvilshah3 reviewed (commented):
Thanks for the PR! I made an initial pass through the changes and left some minor comments.

dhruvilshah3 commented:
@hzxa21 do you have a feel for how much improvement we'd see with only skipping the index sanity check? The reason I ask is because it feels like the lazy mmap change adds some bit of complexity to the code which might be worth adding only if we see a substantial performance benefit with it.

hzxa21 commented:
@dhruvilshah3 I have tested only skipping sanity check. I saw almost no improvement in terms of the broker restart time. The reason is that without lazy mmap change, we do mmap and set the fileds (lastEntry, length, ...) when the index object is created and these operations introduce disk reads to cause prolonged broker restart time. 

The index object is created when we create the segment object during segments loading so only skipping sanity check will not avoid the disk reads and will not help to improve the broker restart time.

lindong28 commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -392,18 +392,27 @@ class Log(@volatile var dir: File,
+          if (timeIndexFileNewlyCreated)
+            segment.timeIndex.resize(0)
+          // Rebuild the index if the index file does not exist
+          if (!offsetIndexExists) {
```
If we want to recover segment when offset index does not exist (which I think is a safe choice), maybe we should also recover segment when time index does not exist for the same reason and for consistency.

hzxa21 replied:
Done.

lindong28 reviewed (commented):
Thanks for the patch @hzxa21.

ijuma commented:
Improving start-up time is definitely welcome. I didn't see a discussion on how errors are handled in this approach. When exactly do we do the `mmap` and how do we propagate the possible errors when we do that?

hzxa21 commented:
@ijuma Thanks for pointing this out. Here is some information about the `mmap` and the error handling:
- `mmap` will be done whenever the index is accessed rather than the index object gets created. To be more specific, `mmap` will be triggered in the following sceanrios:
  1. Sanity check. After this patch, we only sanity check indexes of segments above the recovery points.
   2. Index lookup. This can happen when consumer consumes messages, GroupMetadataManager loads offsets and the segment is truncated/deleted.
  3. Access last entry of the index. The retention thread will acess the last entry of time index in order to determine the max timestamp of a segment for time-based retention.

- Since we don't provide guarantee for the flushed segments, errors (IOException) happened during loading the index file will be catured and the corresponding log dir will be marked as offline. This patch will not change this behavior because the code paths that can trigger index lazy load mentioned above are already wrapped by the `maybeHandleIOException`, which basically catches the IOException, mark the log dir to be offline and throw a KafkaStorageException.

ryannedolan commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -107,10 +128,14 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
+
+  @volatile var initialized = false
+
+  private def initializeMmapAndLength: Unit = {
```
this is a side-effecting method, so should have parens -- `initializeMmapAndLength()`

hzxa21 replied:
Fixed.

ryannedolan commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -107,10 +128,14 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
+
+  private def initializeMmapAndLength: Unit = {
+    val newlyCreated = _file.createNewFile()
+    val raf = if (writable) new RandomAccessFile(_file, "rw") else new RandomAccessFile(_file, "r")
```
more DRY to say:

    val raf = new RandomAccessFile(_file, if (writable) "rw" else "r")

hzxa21 replied:
Fixed.

ryannedolan commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -184,9 +248,14 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
-          mmap = raf.getChannel().map(FileChannel.MapMode.READ_WRITE, 0, roundedNewSize)
-          _maxEntries = mmap.limit() / entrySize
+          _length = Some(roundedNewSize)
+          _mmap = {
```
same here -- more DRY to say:

    _mmap = raf.getChanne.map(if (writable) READ_WRITE else READ_ONLY, 0, roundedNewSize)

hzxa21 replied:
Fixed.

ryannedolan commented on core/src/main/scala/kafka/log/AbstractIndex.scala:
```diff
@@ -120,45 +145,84 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
 
       /* memory-map the file */
-      _length = raf.length()
+      _length = Some(raf.length())
```
Prefer no parens on length method: `Some(raf.length)`

hzxa21 replied:
Fixed.

lindong28 commented:
@hzxa21 Thanks for the update. LGTM. Hey @junrao @ijuma, do you have any concern with this patch?

dhruvilshah3 commented:
One of the things I was wondering is if the change could be simplified by maintaining `indicesOpen` state at the `LogSegment` layer. The first operation that needs to read from or write into any one of the indices could then trigger all of them to be opened. We could probably keep all the logic out of the index layers this way and instead have it all be in one place, in `LogSegment`.

I didn't spend too much time thinking through how much change something like this would require but I wanted to bring it up anyway. Just a suggestion to see if the changes can be simplified.

hzxa21 commented:
@dhruvilshah3 Thanks for the comments. I get your ideas but I wonder whether it actually simplifies this patch. Currently, the memory map operation and the initialization of all the fields of a index object are done when the index object is created. Even though we can maintain a flag in `LogSegment` layer, we still need to add similar logic in the index classes to allow on-demand memory map and initialization. What do you think?

I think the logic in this PR is understandable and not very complicated. Also, this patch is already been tested in our production environment for almost two months and no issues have been seen so we have some confidence in terms of correctness and performance for this PR.

dhruvilshah3 commented:
@hzxa21 I was thinking if we could delay initialization of the indices altogether until an operation that requires us to initialize them. We then don't need any special logic in the index layers.

I think we should merge this PR regardless of whether we can do the above or not, so feel free to ignore the suggestion if you think something like that is not possible without adding more complexity.

ijuma commented:
I agree with @dhruvilshah3's point that it's not great to have to instrument every index method to ensure initialization happens. It's easier to do it one layer above by simply instrumenting the entry point.

hzxa21 commented:
@ijuma I see what you mean. That is a good point. We can have one layer above to trigger the mmap and field initialization all at once instead. Will update the PR achieve that. Thanks for the comments.

ijuma commented:
Sounds good, thanks @hzxa21. This is a very good change btw. :)

lindong28 commented:
I just realized that, based on the experience learned at LinkedIn, some other issues need to be addressed before we commit this patch in Apache Kafka. Here is the history:

1) LinkedIn has very large Kafka cluster and rolling bounce time is too long. So we have hotfix patch as specified in this PR to reduce the time to restart any given broker.

2) The rolling bounce time is significantly reduced, which increases the chance a controller will miss a broker stop/start event. So we have another hotfix patch to fix this scenario. Basically this hotfix patch will ask controller to stop and start a broker if this controller thinks there may be a quick restart.

3) The hotfix patch in 2) make it possible for a broker to receive StopReplicaRequest before the broker receives LeaderAndIsrRequest for all its partitions. With current Kafka implementation, broker will not start ReplicaFetcherThread for the given partition if there is StopReplicaRequest followed by LeaderAndIsrRequest for this partition. In order to address this problem, we have another hotfix patch so that StopReplicaRequest will remove partition `ReplicaManager.allPartitions()`.

4) Due to hotfix patch in 3), if a broker checkpoint highwatermark after receiving StopReplicaRequest, the highwatermark for these partitions will not be included in the checkpoint file and broker will then truncate all these partitions after restart. So we have another hotfix patch as described in KAFKA-6604 to adddres this issue. Note that this issue may also happen if broker receives an outdated LeaderAndIsrRequest that is meant to be delivered before broker restart, as described in KAFKA-6604.

So several issues need to be addressed first before we commit this PR to optimize the broker restart time for Apache Kafka users. In particular, we need to fix the issue described in 2) and https://issues.apache.org/jira/browse/KAFKA-7235.

hzxa21 commented:
@lindong28 Thanks for sharing the context. In my opinion, the current PR address 1) and all the other issues can be addressed by adding broker generation because:
1. For 2), what we need is essentially a mechanism to detect bounced brokers. If we have broker generation, controller can check it in BrokerChange event and implement the fix without the trick we used in the hotfix patch. Note that this is a bug that can potentially happen with and without this PR.
2. For 3) and 4), what we need is essentially a mechnism to let broker reject stall controller request. If we have broker generation, controller can check it when handling StopReplica/LeaderAndIsr/UpdateMetadata and reject requests sent to the previous generation.

What https://issues.apache.org/jira/browse/KAFKA-7235 suggests is adding broker generation support. This requires a KIP and I am currently working on it. Will submit a KIP soon. We can revisit this PR after we have broker generation support.

ijuma commented:
@lindong28 and @hzxa21, thanks for the context on other edge cases that become more common after this pull request. Let's try to get this and KAFKA-7235 (or whatever we decide is best to handle the mentioned edge cases) merged soon.

ijuma commented:
KAFKA-7235 has been merged. Does that mean that this is unblocked or is there more to be done?

hzxa21 commented:
@ijuma Yes. I am doing the rebase and will update this PR when I finish.

ijuma commented:
@hzxa21 Great!

hzxa21 committed:
checkpoint

hzxa21 committed:
KAFKA-7283: Enable mmap indexes lazily and skip sanity check for segments below recovery point to boost broker start time

hzxa21 commented:
@ijuma I have updated the PR to have a thin wrapper layer on top of TimeIndex and OffsetIndex to achieve the lazy initialization without adding complexity in the index layer. I think it is cleaner and easier to understand to do it this way. Can you take a look again?

ijuma commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -65,24 +65,18 @@ class LogSegment private[log] (val log: FileRecords,
     size > rollParams.maxSegmentBytes - rollParams.messagesSize ||
       (size > 0 && reachedRollMs) ||
-      offsetIndex.isFull || timeIndex.isFull || !canConvertToRelativeOffset(rollParams.maxOffsetInMessages)
+      offsetIndex.get.isFull || timeIndex.get.isFull || !canConvertToRelativeOffset(rollParams.maxOffsetInMessages)
```
You can introduce a `offsetIndex` method that calls `get` on the field so that most code stays the same.

hzxa21 replied:
Good point. Thanks. Done.

ijuma commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -94,8 +88,20 @@ class LogSegment private[log] (val log: FileRecords,
+    if (_offsetOfMaxTimestamp < 0) {
+      _offsetOfMaxTimestamp = timeIndex.get.lastEntry.offset
+    }
+    _offsetOfMaxTimestamp
```
Why not handle this in the `Getter` class?

hzxa21 replied:
The reason is that these fields will be updated outside of the index (e.g. https://github.com/apache/kafka/pull/5498/files/44c2d970169f2b85c26af50ecfab4796802abb1a#diff-ac167eb8ca7cc10b6b8fa5772fa438f6R148). If we always access the index to get back these fields, the value can be stale because the index will only get updated when `bytesSinceLastIndexEntry > indexIntervalBytes`. Even if the stale value is fine (but this is a correctness issue),  we don't want to re-read the last entry of the index after the append just in order to get back what we have already known.

ijuma commented on core/src/main/scala/kafka/log/OffsetIndex.scala:
```diff
@@ -197,3 +197,37 @@ class OffsetIndex(_file: File, baseOffset: Long, maxIndexSize: Int = -1, writabl
+  * for the the broker with a lot of log segments
+  *
+  */
+class OffsetIndexGetter(@volatile private var _file: File, baseOffset: Long, maxIndexSize: Int = -1, writable: Boolean = true) {
```
Maybe this could be called `LazyOffsetIndex`.

hzxa21 replied:
Done.

ijuma reviewed (commented):
Thanks for the updated PR. Some initial comments/questions.

hzxa21 committed:
Address comments

hzxa21 commented:
@ijuma Thanks for the prompt review. Updated the PR to address your comments.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -433,18 +433,27 @@ class Log(@volatile var dir: File,
         val baseOffset = offsetFromFile(file)
         val timeIndexFileNewlyCreated = !Log.timeIndexFile(dir, baseOffset).exists()
+        val offsetIndexExists = Log.offsetIndexFile(dir, baseOffset).exists()
+        val timeIndexExists = Log.timeIndexFile(dir, baseOffset).exists()
```
Hmm, do we need timeIndexExists given timeIndexFileNewlyCreated?

hzxa21 replied:
We don't. My bad. Messed up after the rebase.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -433,18 +433,27 @@ class Log(@volatile var dir: File,
+          if (timeIndexFileNewlyCreated)
+            segment.timeIndex.get.resize(0)
+          // Rebuild the index if the index file does not exist
+          if (!offsetIndexExists || !timeIndexExists) {
```
When a user upgrades the message format from pre 0.10 to post 0.10, existing segments will all be missing the time index. Our decision then was not to rebuild the time index for those segments so that we don't increase the broker restart time. That seems to be a reasonable decision. Are you suggesting to change that?

hzxa21 replied:
That is a good point. I miss the point that we have no time index in older versions. I will change it to make it consistent with the previous logic.

junrao commented on core/src/main/scala/kafka/log/Log.scala:
```diff
@@ -433,18 +433,27 @@ class Log(@volatile var dir: File,
             recoverSegment(segment)
+          } else if (baseOffset >= recoveryPoint)
+          // Only sanity check segments above the recovery point
+            segment.sanityCheck()
```
Hmm, not sure that we need to do this. After loadSegments(), we call recoverLog() which already recovers every segment after the recovery point (which forces index rebuild).

hzxa21 replied:
You are right. recoverLog() already recovers every segments after the flushed point so there is no need to do sanity check here. Actually, we no longer need to sanity check the time index and the offset index in any case.

junrao commented on core/src/main/scala/kafka/log/TimeIndex.scala:
```diff
@@ -220,3 +220,37 @@ class TimeIndex(_file: File, baseOffset: Long, maxIndexSize: Int = -1, writable:
+  *
+  */
+class TimeIndexGetter(@volatile private var _file: File, baseOffset: Long, maxIndexSize: Int = -1, writable: Boolean = true) {
+  private var timeIndex: Option[TimeIndex] = None
```
Should this be volatile since it can be set and read from different threads?

hzxa21 replied:
Yes. Done.

junrao commented on core/src/main/scala/kafka/log/OffsetIndex.scala:
```diff
@@ -197,3 +197,37 @@ class OffsetIndex(_file: File, baseOffset: Long, maxIndexSize: Int = -1, writabl
+  *
+  */
+class OffsetIndexGetter(@volatile private var _file: File, baseOffset: Long, maxIndexSize: Int = -1, writable: Boolean = true) {
+  private var offsetIndex: Option[OffsetIndex] = None
```
Should this be volatile since it can be set and read from different threads?

hzxa21 replied:
Yes. Done.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -94,8 +88,20 @@ class LogSegment private[log] (val log: FileRecords,
+  def offsetOfMaxTimestamp: Long = {
+    if (_offsetOfMaxTimestamp < 0) {
+      _offsetOfMaxTimestamp = timeIndex.get.lastEntry.offset
+    }
```
No need for {} for single line statement.

hzxa21 replied:
Done.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -94,8 +88,20 @@ class LogSegment private[log] (val log: FileRecords,
+  def maxTimestampSoFar: Long = {
+    if (_maxTimestampSoFar < 0) {
+      _maxTimestampSoFar = timeIndex.get.lastEntry.timestamp
+    }
```
No need for {} for single line statement.

hzxa21 replied:
Done.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -94,8 +88,20 @@ class LogSegment private[log] (val log: FileRecords,
   /* The maximum timestamp we see so far */
-  @volatile private var maxTimestampSoFar: Long = timeIndex.lastEntry.timestamp
-  @volatile private var offsetOfMaxTimestamp: Long = timeIndex.lastEntry.offset
+  @volatile private var _maxTimestampSoFar: Long = -1L
```
When there is no time index (old message format), the default timestamp is -1. So, we will need a different default value.

hzxa21 replied:
I see. I will use Option[Long] instead.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -65,24 +65,18 @@ class LogSegment private[log] (val log: FileRecords,
-    }
-    else throw new NoSuchFileException(s"Offset index file ${offsetIndex.file.getAbsolutePath} does not exist")
+  def sanityCheck(): Unit = {
+    offsetIndex.get.sanityCheck()
```
Should we still make those sanity checks when the indexes are (lazily) created?

hzxa21 replied:
We don't need to sanity check offse index and the time index anymore.

ijuma replied:
Can you please explain why?

ijuma replied:
Also, let's include the explanation in the PR description.

hzxa21 replied:
@ijuma The purpose of the sanity check is to detect index corruption (if any) and recover the index. Since in recoverLog() we will recovers every segments after the flushed point in any case,  calling offsetIndex.sanityCheck() and timeIndex.sanityCheck() are redundant operations.

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -3722,6 +3731,8 @@ object LogTest {
       segment.append(MemoryRecords.withRecords(baseOffset + Int.MaxValue - 1, CompressionType.NONE, 0,
         record(baseOffset + Int.MaxValue - 1)))
+      Log.offsetIndexFile(logDir, baseOffset).createNewFile()
+      Log.timeIndexFile(logDir, baseOffset).createNewFile()
```
Do we need to explicitly create these 2 index files?

hzxa21 replied:
We need. Otherwise `recoverSegment()` will be invoked because the offset index is missing and truncation will happen so that the assertion here will fail: https://github.com/apache/kafka/blob/trunk/core/src/test/scala/unit/kafka/log/LogTest.scala#L3690

Btw, this happened to work without this patch because the index file will always get created before the offset index file existence check (See: https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/AbstractIndex.scala#L112), which means we will never reach https://github.com/apache/kafka/blob/trunk/core/src/main/scala/kafka/log/Log.scala#L475 before this patch.

junrao replied:
Ok, Probably add a comment why we need to explicitly create those files.

hzxa21 replied:
Done.

junrao reviewed (commented):
@hzxa21 : Thanks for the patch. A few comments below. Also, I don't see the changes addressing Ismael's comments.

hzxa21 committed:
Address comments

hzxa21 commented:
@junrao @ijuma Thanks for the comments. I push to the wrong remote branch previously so the updates didn't show up in this PR. I have already fixed that and updated the PR again to address all your comments.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -92,19 +96,20 @@ class LogSegment private[log] (val log: FileRecords,
-  @volatile private var _offsetOfMaxTimestamp: Long = -1L
+
+  @volatile private var _offsetOfMaxTimestamp: Option[Long] = None
+  def offsetOfMaxTimestamp_=(offset: Long): Unit = _offsetOfMaxTimestamp = Some(offset)
```
Naming wise, should we postfix this with SoFar to be consistent with maxTimestampSoFar?

hzxa21 replied:
Done.

junrao commented on core/src/main/scala/kafka/log/LogSegment.scala:
```diff
@@ -77,10 +77,14 @@ class LogSegment private[log] (val log: FileRecords,
-    txnIndex.sanityCheck()
+  def sanityCheck(timeIndexFileNewlyCreated: Boolean): Unit = {
+    if (lazyOffsetIndex.file.exists) {
+      // Resize the time index file to 0 if it is newly created.
```
Perhaps add a comment why we skip the sanity check just for offsetIndex and timeIndex?

hzxa21 replied:
Done.

ijuma commented:
Thanks @hzxa21. Can you please rebase?

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1798,7 +1798,7 @@ class LogTest {
 
     // reopen the log
-    log = createLog(logDir, logConfig, recoveryPoint = 200L)
+    log = createLog(logDir, logConfig)
```
Is this change intended? The default value for  recoveryPoint is not 200.

hzxa21 replied:
Yes. After we get rid of the sanity check, the index files for segment below the recovery point will no longer get checked on startup so in order to test rebuild a corrupted index, we need to make sure the recovery point is below the base offset for the corrupted index.

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -1911,10 +1911,17 @@ class LogTest {
     val bogusIndex2 = Log.offsetIndexFile(logDir, 5)
     val bogusTimeIndex2 = Log.timeIndexFile(logDir, 5)
 
+    // Create index files
```
Probably add a comment why we need to explicitly create those files.

hzxa21 replied:
Done.

junrao commented on core/src/test/scala/unit/kafka/log/LogTest.scala:
```diff
@@ -2363,6 +2370,8 @@ class LogTest {
   private def testDegenerateSplitSegmentWithOverflow(segmentBaseOffset: Long, records: List[MemoryRecords]): Unit = {
     val segment = LogTest.rawSegment(logDir, segmentBaseOffset)
+    Log.offsetIndexFile(logDir, segmentBaseOffset).createNewFile()
+    Log.timeIndexFile(logDir, segmentBaseOffset).createNewFile()
```
Probably add a comment why we need to explicitly create those files.

hzxa21 replied:
Done.

junrao reviewed (commented):
@hzxa21 : Thanks for the updated patch. Just a few minor comments below.

hzxa21 committed:
Address comments

junrao commented:
retest this please

junrao reviewed (approved):
@hzxa21 : Thanks for the patch. LGTM. I am not sure why the tests failed. Triggering a retest.

hzxa21 commented:
@junrao Thanks for the review. The tests are passed now.

junrao commented:
@hzxa21 : Just a followup question. I am not sure if there is an existing test that exercises the logic of loading the offset/timestamp index on demand. If not, could you add one?

hzxa21 commented:
@junrao Some tests implicitly exercise the logic but I do think it also makes sense to add a separate unit test to test this specifically. I will open a follow-up PR for that.

junrao commented:
@hzxa21 : Thanks.
